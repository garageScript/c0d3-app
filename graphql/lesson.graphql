query Lesson {
  lessons {
    id
    title
  }
}