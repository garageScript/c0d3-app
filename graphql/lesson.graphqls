type Lesson {
  id: String
  title: String
}

type Query {
  lessons: Lesson!
}