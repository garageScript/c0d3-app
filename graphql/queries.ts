import { gql } from 'apollo-boost'

export const GET_LESSONS = gql`
  query Lessons {
    curriculumStatus {
      id
      description
      title
      order
      docUrl
      challenges {
        id
      }
      currentUser {
        userLesson {
          isEnrolled
          isTeaching
          isPassed
        }
      }
    }
    lessons {
      id
      title
      description
      order
      docUrl
      challenges {
        id
        description
        title
        order
      }
    }
  }
`

export const GET_LESSON = gql`
  query lessonInfo(
    $lessonInfo: LessonId
    $lessonUserInfo: LessonUserId
    $lessonStatus: LessonId
  ) {
    lessonInfo(input: $lessonInfo) {
      id
      title
      description
      order
      docUrl
      chatUrl
      challenges {
        id
        description
        title
        order
      }
    }
    userSubmissions(input: $lessonUserInfo) {
      id
      status
      mrUrl
      diff
      viewCount
      comment
      challengeId
      reviewer {
        id
        username
      }
      createdAt
      updatedAt
    }
    lessonStatus(input: $lessonStatus) {
      isPassed
      isTeaching
      isEnrolled
    }
  }
`

export const GET_SUBMISSIONS = gql`
  query submissions($in: LessonId) {
    submissions(input: $in) {
      status
    }
  }
`
