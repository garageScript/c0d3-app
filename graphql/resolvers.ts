import { login, logout, signup } from '../helpers/controllers/authController'
import db from '../helpers/dbload'

const { Lesson, User } = db

export default {
  Query: {
    lessons() {
      return Lesson.findAll({
        include: [
          'challenges',
          {
            model: User,
            through: {
              attributes: ['isPassed', 'isTeaching', 'isEnrolled']
            }
          }
        ],
        order: [['order', 'ASC']]
      })
    }
  },

  Mutation: {
    login,
    logout,
    signup
  }
}
