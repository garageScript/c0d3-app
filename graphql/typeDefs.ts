import { gql } from 'apollo-boost'

export default gql`
  type Query {
    lessons: [Lesson]
  }

  type Mutation {
    login(username: String!, password: String!): AuthResponse
    logout: AuthResponse
    signup(
      firstName: String!
      lastName: String!
      email: String!
      username: String!
      password: String!
    ): AuthResponse
    deleteUser(userId: String!): AuthResponse
  }

  type AuthResponse {
    success: Boolean
    username: String
    error: String
  }

  type Submission {
    id: String
    status: String
    mrUrl: String
    diff: String
    viewCount: Int
    comment: String
    userId: String
    order: Int
    lessonId: String
    challengeId: String
    challenge: Challenge
    reviewer: User
    user: User
    reviewerId: String
    createdAt: String
    updatedAt: String
  }

  type User {
    id: String
    username: String
    userLesson: UserLesson
    email: String
    name: String
    isAdmin: Boolean
  }

  type UserLesson {
    id: String
    userId: String
    lessonId: String
    isPassed: String
    isTeaching: String
    isEnrolled: String
    starGiven: User
    starComment: String
  }

  type Lesson {
    id: String
    description: String
    docUrl: String
    githubUrl: String
    videoUrl: String
    order: Int
    title: String
    challenges: [Challenge]
    users: [User]
    currentUser: User
    chatUrl: String
  }

  type Challenge {
    id: String
    description: String
    lessonId: String
    title: String
    order: Int
  }
`
