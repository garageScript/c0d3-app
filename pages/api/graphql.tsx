import { ApolloServer, gql } from 'apollo-server-micro'

const typeDefs = gql`
  type Query {
    username: String
    hello: String
  }
`

const resolvers = {
  Query: {
    username(_parent: any, _args: any, ctx: any) {
      return ctx.user
    },
    hello() {
      return 'hello'
    }
  }
}

const context = ({ req }: { req: any }) => req.cookies

const apolloServer = new ApolloServer({ typeDefs, resolvers, context })

export const config = {
  api: {
    bodyParser: false
  }
}

export default apolloServer.createHandler({ path: '/api/graphql' })
