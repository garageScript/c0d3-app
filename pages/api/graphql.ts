import { ApolloServer } from 'apollo-server-micro'
import session from 'express-session'
import nextConnect from 'next-connect'
import db from '../../helpers/dbload'
import typeDefs from '../../graphql/typeDefs'
import resolvers from '../../graphql/resolvers'
import connectSequelize from 'connect-session-sequelize'

// Sequelize Store for express-sessions
const { sequelize } = db
const SequelizeStore = connectSequelize(session.Store)

const ONE_WEEK = 1000 * 60 * 60 * 24 * 7

const handler: any = nextConnect() // For session middleware. TODO: Need to define types for Session
const apolloServer = new ApolloServer({
  typeDefs,
  resolvers,
  context: ({ req }) => ({ req }) // This lets GraphQL have access to sessions
})

const graphQLHandler = apolloServer.createHandler({ path: '/api/graphql' })

handler
  .use(
    session({
      secret: process.env.SESSION_SECRET || '',
      store: new SequelizeStore({
        db: sequelize
      }),
      resave: false,
      saveUninitialized: false,
      cookie: {
        domain: process.env.DB_HOST,
        maxAge: ONE_WEEK
      }
    })
  )
  .get('/api/graphql', graphQLHandler)
  .post('/api/graphql', graphQLHandler)

export const config = {
  api: {
    bodyParser: false // This is required for GraphQL to work properly  https://blog.logrocket.com/building-a-graphql-server-in-next-js/
  }
}

export default handler
