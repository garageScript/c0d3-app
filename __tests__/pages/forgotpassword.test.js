import React from 'react'
import { GraphQLError } from 'graphql'
import { render, fireEvent, waitFor, act } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import GET_APP from '../../graphql/queries/getApp'
import RESET_PASSWORD from '../../graphql/queries/resetPassword'
import dummySessionData from '../../__dummy__/sessionData'
import ForgotPassword from '../../pages/forgotpassword'

describe('ForgotPassword Page', () => {
  test('Should render password reset instructions on success', async () => {
    const validUser = 'fakevaliduser'

    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session: dummySessionData,
            lessons: [],
            alerts: []
          }
        }
      },
      {
        request: {
          query: RESET_PASSWORD,
          variables: { userOrEmail: validUser }
        },
        result: {
          data: {
            reqPwReset: {
              success: true,
              token: 'faketoken'
            }
          }
        }
      }
    ]

    const { container, getByTestId, getByRole } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ForgotPassword />
      </MockedProvider>
    )

    const submitButton = getByTestId('submit')
    const userOrEmail = getByTestId('userOrEmail')

    // wait for formik validations that are async
    fireEvent.change(userOrEmail, {
      target: {
        value: validUser
      }
    })

    await waitFor(() => fireEvent.click(submitButton))

    await waitFor(() =>
      getByRole('heading', { name: /password reset instructions sent/i })
    )

    // wait for mutation updates after submitButton is clicked
    await waitFor(() => expect(container).toMatchSnapshot())
  })

  test('Should render invalid user/email form on error', async () => {
    const invalidUser = 'fakeinvaliduser'

    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session: dummySessionData,
            lessons: [],
            alerts: []
          }
        }
      },
      {
        request: {
          query: RESET_PASSWORD,
          variables: { userOrEmail: invalidUser }
        },
        result: {
          errors: [new GraphQLError('invalid user')]
        }
      }
    ]

    const { container, getByTestId, getByRole } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ForgotPassword />
      </MockedProvider>
    )

    const submitButton = getByTestId('submit')
    const userOrEmail = getByTestId('userOrEmail')

    // wait for formik validations that are async
    fireEvent.change(userOrEmail, {
      target: {
        value: invalidUser
      }
    })
    await waitFor(() => fireEvent.click(submitButton))

    await waitFor(() =>
      getByRole('heading', { name: /username or email does not exist/i })
    )

    // wait for mutation updates after submitButton is clicked
    await waitFor(() => expect(container).toMatchSnapshot())
  })
})
