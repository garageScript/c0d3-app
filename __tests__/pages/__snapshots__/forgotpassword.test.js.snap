// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword Page Should render invalid user/email form on error 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="/docs/learn"
            >
              Learn
            </a>
            <a
              class="nav-item nav-link"
              href="https://discord.gg/c0d3"
              rel="noopener noreferrer"
              target="_blank"
            >
              Community
            </a>
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="dropdown-toggle btn btn-none"
                type="button"
              >
                Admin
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="d-flex"
      >
        <a
          class="btn btn-secondary border overflow-hidden p-2 text-truncate"
          href="/profile/fakeusername"
        >
           fakeusername
        </a>
        <button
          class="btn border ml-2"
          style="color: rgb(41, 41, 41);"
        >
          Logout
        </button>
      </div>
    </div>
  </nav>
  <div
    class="container-md"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <h1
            class="card-title h2 font-weight-bold mb-5"
          >
            Username or Email does not exist
          </h1>
          <p
            class="card-text"
          />
          <a
            class="btn btn-primary btn-lg btn-block mb-3"
            href="/forgotpassword"
            role="button"
          >
            Go Back
          </a>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2021 GarageScript
  </footer>
</div>
`;

exports[`ForgotPassword Page Should render password reset instructions on success 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="/docs/learn"
            >
              Learn
            </a>
            <a
              class="nav-item nav-link"
              href="https://discord.gg/c0d3"
              rel="noopener noreferrer"
              target="_blank"
            >
              Community
            </a>
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="dropdown-toggle btn btn-none"
                type="button"
              >
                Admin
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="d-flex"
      >
        <a
          class="btn btn-secondary border overflow-hidden p-2 text-truncate"
          href="/profile/fakeusername"
        >
           fakeusername
        </a>
        <button
          class="btn border ml-2"
          style="color: rgb(41, 41, 41);"
        >
          Logout
        </button>
      </div>
    </div>
  </nav>
  <div
    class="container-md"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <img
            alt="Green circle with a checkmark inside."
            class="mb-4"
            height="100px"
            src="/assets/curriculum/icons/checked.svg"
            width="100px"
          />
          <h1
            class="card-title h2 font-weight-bold mb-5"
          >
            Password reset instructions sent
          </h1>
          <p
            class="card-text"
          />
          <p>
            You will receive an email containing a link to reset your password.
          </p>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2021 GarageScript
  </footer>
</div>
`;
